guilib/measure/classes/abc.py

```diff
@@ -0,0 +1,603 @@
+            return False
+        invalid = self.are_settings_invalid(new_settings)
+        for missing, *info in invalid:
+            emit_error(self, QObjectSettingsErrors.INVALID_SETTINGS,
+                       missing, ' '.join(info))
+            return False
+
+        self._settings = new_settings
+        return True
+
+
+# DISABLE: seems a bug in pylint. This is also an ABC.
+# pylint: disable-next=abstract-method
+class HardwareABC(QObjectWithSettingsABC):
+    """Abstract base class of hardware-related objects."""
+
+    # Emitted whenever the busy state of the device changes.
+    # Contains the new busy state of the device.
+    busy_changed = qtc.pyqtSignal(bool)
+
+    def __init__(self, *args, **kwargs):
+        """Initialise instance."""
+        # Busy state of the device.
+        self._busy = False
+        super().__init__(*args, **kwargs)
+
+    @property
+    def busy(self):
+        """Return busy state of instance."""
+        return self.get_busy()
+
+    @busy.setter
+    def busy(self, is_busy):
+        """Set busy state of instance.
+
+        Note that 'self.busy =' will be performed in the calling
+        thread. To avoid this, one can invoke the set_busy slot.
+
+        Parameters
+        ----------
+        is_busy : bool
+            True if the instance is busy.
+
+        Emits
+        -----
+        busy_changed
+            If the busy state of the instance changed.
+        """
+        self.set_busy(is_busy)
+
+    def get_busy(self):
+        """Return whether the instance is busy.
+
+        Note that this method is called in the getter of the busy
+        property. If this method is overridden, this will also
+        change the value that is returned by the busy property.
+
+        Returns
+        -------
+        busy : bool
+            True if the instance is busy.
+        """
+        return self._busy
+
+    @qtc.pyqtSlot(bool)
+    def set_busy(self, is_busy):
+        """Set busy state of the instance.
+
+        Parameters
+        ----------
+        is_busy : bool
+            True if the instance is busy.
+
+        Emits
+        -----
+        busy_changed
+            If the busy state of the instance changed.
+        """
+        was_busy = self.busy
+        is_busy = bool(is_busy)
+        self._busy = is_busy
+        if was_busy != is_busy:
+            self.busy_changed.emit(self.busy)
+
+
+class DeviceABC(HardwareABC):
+    """Abstract base class of hardware device objects."""
+
+    @abstractmethod
+    def list_devices(self):
+        """List all devices of this class.
+
+        This method must return a list of SettingsInfo instances. Each
+        device is represented by a single SettingsInfo instance. The
+        SettingsInfo object must contain a .unique_name and can contain
+        .more information as a dict. The information contained within
+        a SettingsInfo must be enough to determine a suitable settings
+        file for the device from it.
```

Comment:
Perhaps we should be more explicit as to what "suitable" means? It is a bit generic.